src/graph/type_indexededgelist.c

```diff
@@ -373,6 +415,37 @@ igraph_error_t igraph_add_vertices(igraph_t *graph, igraph_integer_t nv, void *a
         }
     }
 
+    /* modification successful, clear the cached properties of the graph.
+     *
+     * Adding one or more nodes does not change the following cached properties:
+     *
+     * - IGRAPH_PROP_HAS_LOOP
+     * - IGRAPH_PROP_HAS_MULTI
+     * - IGRAPH_PROP_HAS_RECIPROCAL
+     * - IGRAPH_PROP_IS_DAG (adding a node does not create/destroy cycles)
+     * - IGRAPH_PROP_IS_FOREST (same)
+     *
+     * Adding one or more nodes without any edges incident on them is sure to
+     * make the graph disconnected (weakly or strongly), so we can keep the
+     * connectivity-related properties if they are currently cached as false.
+     * (Actually, even if they weren't cached as false, we could still set them
+     * to false, but we don't have that functionality yet).
+     */
```

Comment:
There is a special case when `is_connected` is not `keep_when_false`.

If the graph has no vertices, we consider it disconnected. Adding a single vertex to it will make it connected.

To keep things as simple as possible, we can invalidate the cache unconditionally when the final vertex count is less than 2. There is little gain from caching properties for the 0- or 1-vertex graph.

Reply:
Aaaargh, these are _exactly_ the special cases because of which I'm scared to add any smart logic to cached properties. I'll fix this.